### ISSUE TITLE
VBET Memory crash for large huc 10190003

### ISSUE THREAD

**Issue by MattReimer** [Riverscapes/riverscapes-tools#191]
Let's see how much more memory this needs. if it's simply about bumping the docker box a bit as a stop-gap then we can do it.

```
2021-01-28 03:03:47::         PROGRESS: 94 / 100  (94%)  Polygonizing raster     (Ellapsed: 1:70:50)
2021-01-28 03:05:05::         PROGRESS: 95 / 100  (95%)  Polygonizing raster     (Ellapsed: 1:72:09)
2021-01-28 03:06:15::         PROGRESS: 96 / 100  (96%)  Polygonizing raster     (Ellapsed: 1:73:19)
2021-01-28 03:07:25::         PROGRESS: 97 / 100  (97%)  Polygonizing raster     (Ellapsed: 1:74:28)
2021-01-28 03:08:23::         PROGRESS: 98 / 100  (98%)  Polygonizing raster     (Ellapsed: 1:75:27)
2021-01-28 03:08:59::         PROGRESS: 99 / 100  (99%)  Polygonizing raster     (Ellapsed: 1:76:03)
2021-01-28 03:21:07::         PROGRESS: 100 / 100  (100%)  Polygonizing raster     (Ellapsed: 1:88:10)
2021-01-28 03:21:07:: [info] [Polygonizing raster] Completed 100 operations.  Total Time: 1:88:10     
2021-01-28 03:21:07:: [debug] [DatasetRegistry] Dataset closed: /usr/local/data/vbet/10190003/output/intermediates/vbet_intermediates.gpkg
2021-01-28 03:21:07:: [info] [VBET] Done
2021-01-28 03:21:07:: [info] [VBET] Sanitizing
2021-01-28 03:21:07:: [info] [VectorBase] Dataset not found. Creating: /usr/local/data/vbet/10190003/output/outputs/vbet.gpkg
2021-01-28 03:21:07:: [debug] [VectorBase] Dataset created: /usr/local/data/vbet/10190003/output/outputs/vbet.gpkg
2021-01-28 03:21:07:: [debug] [VectorBase] No layer named "vbet_50" found
2021-01-28 03:21:07:: [debug] [DatasetRegistry] Dataset opened: /usr/local/data/vbet/10190003/output/intermediates/vbet_intermediates.gpkg for READING
2021-01-28 03:21:07::         PROGRESS: 1 / 46,059  (0%)  Sanitizing     (Ellapsed: 0.0 seconds)
2021-01-28 03:22:13:: /usr/local/runner.sh: line 69:    67 Killed                  vbet $HUC 
```

### Update:

I've implemented rudimentary memory profiling as an additional `--debug` parameter on all our tools. It spits out a seperate logfile and a matplotlib png file like so:

![rs_context_memusage](https://user-images.githubusercontent.com/1063391/106168191-d8ebcd00-6142-11eb-8774-339f20d2f47f.png)

This is going to let us do two things:

1. Identify any spikes that go over the amount of RAM we have in cybercastor
2. Now we can marry the timestamp from a memory spike to figure out where in the code the problem is.

**Comment by MattReimer** [comment]
I kicked off this job as soon as it failed and it's still running. 

As a stop-gap I've increased the RAM on cybercastor so this _should_ finish. That should buy us enough time to make the correct fix.

**Comment by MattReimer** [comment]
New branch for this work. The multi-threaded memory profiler is here:
https://github.com/Riverscapes/riverscapes-tools/tree/memprofiling

**Comment by MattReimer** [comment]
As of this morning we have totally flat memory curves on VBET. 

![vbet_mem_windows](https://user-images.githubusercontent.com/1063391/107099916-c8cd9080-67c7-11eb-98ed-f7844d195285.png)
![vbet_mem_wslUbuntu](https://user-images.githubusercontent.com/1063391/107099920-ca975400-67c7-11eb-9145-04a7bd9f8f1d.png)

Just making sure this is fixed for 10190003 and then we can close this.

**Comment by MattReimer** [comment]
Some rough math shows that we've reduced VBET calculation times by 25% overall but most importantly we don't exceed the memory we're allotted so this is done. Nice job @KellyMWhitehead
